src/Microsoft.DotNet.Helix/JobSender/JobDefinition.cs

```
@@ -202,6 +203,20 @@ public async Task<ISentJob> SendAsync(Action<string> log = null)
             string jobListUriForLogging = jobListUri.ToString().Replace(jobListUri.Query, "");
             log?.Invoke($"Created job list at {jobListUriForLogging}");
 
+            // Only specify the ResultContainerPrefix if both repository name and source branch are available.
+            if (!string.IsNullOrEmpty(Environment.GetEnvironmentVariable("BUILD_REPOSITORY_NAME")) && !string.IsNullOrEmpty(Environment.GetEnvironmentVariable("BUILD_SOURCEBRANCH")))
+            {
+                // Remove all slashes from repository name and branch name. Also remove refs/, heads/, and pulls/ from branch name.
+                // ResultContainerPrefix will be <Repository Name>-<BranchName>
+                string repoName = Environment.GetEnvironmentVariable("BUILD_REPOSITORY_NAME").Replace("/", "-");
+                string branchName = Environment.GetEnvironmentVariable("BUILD_SOURCEBRANCH")
+                    .Replace("refs/","")
```

The tester in me says that this will fail hard for a branch called `refs/heads/heads/refs/pull/5` I'd be so tempted to `git checkout -b heads/refs/pull/5`

---

The documentation says it'll be "refs/heads/...", let's assert it looks like that (fail if it doesn't) and then always just Substring(11)

---

Hrm... or let's just not touch it, since PR's are weird.

---

Yeah, PRs are refs/pull and regular branches are refs/heads/. Maybe we should do something like... excuse the pseudocode:

```
splitBranch = branch.split(/)
if (PR)
{
    branch = splitBranch[1]-splitBranch[2] // this should give us pull-###
}
else
{
    splitBranch.RemoveRange(0,2); // remove refs/heads
    branch = string.Join(splitBranch, "-") // put everything after the first refs/heads back together with dashes
}
```

Thoughts?

---

Does it NEED to be cleaned up?  Is the goal just to get a reliable prefix for the container, leaving the refs-heads doesn't really hurt.

---
In reply to: [270215624](https://github.com/dotnet/arcade/pull/2367#discussion_r270215624) [](ancestors = 270215624)

---

it might not matter, but if the branch name is really long, the prefix could get truncated (we are limited to 45 characters, to support 18 characters of the correlation id). Removing unnecessary bits early makes it automatically smaller, giving us more from the branch name. This would be like, if someone managed to start a job off of their actual branch, rather than from a PR, so maybe we just let those get truncated?

---

The normal use case is either "refs/heads/master", "refs/heads/release/X.0", or "refs/pulls/5000/merge".  The 45 includes the "repo" right?  Hrm... that probably costs us another 15 characters, leaving it down to 30...

refs/heads/ is 10...

So for random branches running builds (I'm not even sure how to actually do that with arcade...), we leave them 20 for the branch... and if it's truncated, it's still a safely unique name, right?

I'd... leave it un-modified for now to simplify it.  If it proves to be a problem later, we can add the complication then, but I don't feel like it's going to be a problem, and the badness isn't that bad.

---
In reply to: [270227509](https://github.com/dotnet/arcade/pull/2367#discussion_r270227509) [](ancestors = 270227509)

---

Sounds reasonable. I think you're right that we'll probably be fine not modifying it. I will update this.

---

Ok, I've fixed this up some. I tested leaving in periods, like in 3.0, but that breaks the storage uploader, so I added a replace to remove them. Are there any other special characters you guys can think of that might be in branch names that I should look out for? Do you think it would be better to break these out into a helper function to sanitize branch and repo names?

---

Ok, it looks like I'm going to have to remove anything that isn't a-z 0-9 or a dash from both repo and branch.

---

There is a rather large set of things that are allowed in branch names but not allowed as part of a container name. It would probably be best to do a regex replace for anything not allowed in a container name. e.g. Replace `[^a-zA-Z0-9-]` with '-', then replace `-*--*` with '-' to remove consecutive dashes.

---

actually, the second regex should be `--+`

---

I actually just removed all characters we don't allow, and replaced / with -, but you're right, we might end up in a situation with double dashes, so I'll go replace multiple dashes with single dashes. Any reason the second regex isn't -+?

---

Go the other direction, only allow characters you know are good.

`Regex.Replace(string, "[^A-Za-z0-9_-]", "-")`

---

"Don't allow" will be a recipe for bugs. I assume emojis are also not allowed.  Always allow list things when doing input validation, not deny list.

---

Right right right, that's what I did. The regex is [^a-z0-9-], which are replaced with an empty string.